Enable Image encoding by leveraging existing Skia functionality

majido opened the pull request:
Enable Image encoding by leveraging existing Skia functionality.

This  patch introduces the following:
- Add a new encodeImage to painting library that takes Image, format, quality parameters.
- Asynchronously Encode underlying SkImage with given format and quality on IO thread.
- Post encoded result to the callback.

At the moment, it only exposes JPG, PNG, WEBP formats since it is easier to add more later
when there is a need.

Given this functionality it becomes easy to use the provided Canvas API in a flutter app to
draw and then export the result to an encoded image for sharing/download etc.

Fixes https://github.com/flutter/flutter/issues/11648

majido commented:
I needed this functionality on my own project and thought this may be of general value since 
there was already an issue filed for it.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
+  JPEG,
+  PNG,
+  WEBP,
+}
```
these will need documentation, ideally including a pointer to the relevant specification for each one.

Hixie replied:
Also for each one we should say what the `quality` argument does.

majido replied:
These map to Skia values. Skia's does not document the supported format in much details. I was not able to even find link to relevant specification from them. So I have not added any spec links since I was not confident. If you have authoritative links I am happy to add.

Here is the Skia [documentation](https://skia.org/user/api/SkImage_Reference#SkImage_encodeToData_quality
) on quality. It has no details on exact impact of quality for each format beyond "it is format specific and may be ignored". 

My tests using the demo in the above link confirms that the quality actually works for this three
formats. So I have updated the wording to reflect this with more detail on low vs high.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
+}
+
+/// Callback signature for [encodeImage].
+typedef void ImageEncoderCallback(Uint8List result);
```
Please document what the "result" argument is (presumably, bytes representing the encoded image in the requested format?).

Hixie replied:
(this can just be removed if we go the route suggested below of returning a future)

majido replied:
Removed. Using a Future instead.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
+/// Callback signature for [encodeImage].
+typedef void ImageEncoderCallback(Uint8List result);
+
+/// Convert an [Image] object into a byte array/
```
typo: / -> .

majido replied:
Done.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
+/// [format] is encoding format to be used.
+///
+/// [quality] is a value in [0, 100] where 0 corresponds to the lowest quality.
+void encodeImage(Image image, EncodedImageFormat format, int quality, ImageEncoderCallback callback)
```
This should be a method on Image.

Hixie replied:
We should probably make the format and quality arguments named optional arguments with defaults.
It's probably also a good idea to have this return a `Future<ByteData>` rather than using the callback approach. That would be more idiomatic. We can use the callback approach internally in this function in order to implement the Future, though, since dealing with Futures and the engine is non-trivial otherwise.

Hixie replied:
Also, if we make this a method on Image, we should probably call it `toByteData`, to be more idiomatic.

tvolkert replied:
+1.  FYI for context, see https://github.com/flutter/flutter/issues/11648#issuecomment-350902741

majido replied:
Ack. I think it makes sense for this to be a method on Image. 

Can you please provide my an example of how a Future returning method is implemented in
engine?
I presume Hixie's suggestion was to have a private method implemented natively that uses
callback and then call this private method from public dart method. Effectively creating
the future in dart and resolving that future via callback roughly:

``` 
class Image {
  Future<Uint8List> toByteData(format, quality) {
      final Completer<Uint8List> completer = new Completer<Uint8List>();
      _encodeImage(this, format, quality, completer.complete); // this is natively implemented.
     return completer.future;
   }
}
```

Hixie replied:
Yup, that's it exactly.

Hixie replied:
You can just use the `_futurize` method in this file to do it for you though.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
+///
+/// [format] is encoding format to be used.
+///
+/// [quality] is a value in [0, 100] where 0 corresponds to the lowest quality.
```
Please start each paragraph with a capital letter, e.g. by saying `The [format] argument is...`

Hixie replied:
Rather than `in [0, 100]` let's say something like `in the range 0 to 100`.

majido replied:
Done. Improved the wording and provide more information of what lowest and highest quality mean.

Hixie commented:
cc @chinmaygarde for review, especially of the C++ side with which I am less familiar.

Fixes https://github.com/flutter/flutter/issues/11648.

Hixie commented:
Thanks for your contribution! This is a super-popular request, so you will definitely receive the gratitude of many for implementing this!

Hixie commented:
It would be great if you could include a test of this (just verify that some trivial Image, created with a Canvas, returns the expected bytes). Other relevant tests are in https://github.com/flutter/engine/blob/master/testing/dart/ .

Hixie commented:
The Travis failure is some minor thing with our license script -- you just need to edit the file as suggested in the Travis output to add your FILE lines.

chinmaygarde commented on lib/ui/painting/image_encoding.cc:
```diff
@@ -0,0 +1,136 @@
+      tonic::DartConverter<CanvasImage*>::FromArguments(args, 0, exception);
+
+  if (exception) {
+    Dart_ThrowException(exception);
```
We are trying to reduce/remove our use of `Dart_ThrowException` because it does not return and hence will not call destructors of C++ objects on the stack (not that this is an issue in your patch). Can you make this method return an error flag and message that is sent back to the Dart method? On error, you can then throw a Dart exception from that wrapper to get similar behavior.

majido replied:
Done. This is no longer uses Dart_ThrowException instead relies on return values.

The error value is handled via existing _futurize method in the dart side and turned into
and exception.

eseidelGoogle commented:
<3 This will make a lot of people happy.  Thank you!

nsreenath commented:
Good work @majido 
Will this also allow us to save a widget tree as PNG ?

tvolkert commented:
@majido just checking in to see if there's anything you need to address the review comments.

majido commented:
PR is now updated. I believe all issues should have been addressed.

I also added a simple test which compares a generated canvas image against its encoded
versions.

majido commented:
@nsreenath 
> Will this also allow us to save a widget tree as PNG ?

This patch allows one to convert any canvas into an encoded image.  My usage is mainly based on directly drawing into a canvas. Something similar to this [example](https://github.com/flutter/engine/blob/master/examples/hello_flutter/lib/main.dart). So if there is a way to paint the widget tree into a canvas, then perhaps yes.

I haven't look at the internals of flutter rendering system, but each render object seems to have 
a `paint` function. So there maybe a way to trigger a paint on the widget tree root and obtain
the resulting canvas and then save that. The recursive paint appears to be creating new canvases for
each layer so there is definitely more complexities I am overlooking. Hopefully real experts here may be able to shed more light.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1196,6 +1196,16 @@ class Paint {
+  JPEG,
+  PNG,
+  WEBP,
+}
```
These should be lowercase.

Here are some doc strings for each one:
```
/// JPEG format.
///
/// This format, strictly speaking called JFIF, is a lossy compression graphics format that
/// can handle images up to 65,535 pixels in either dimension. The `quality` metric controls
/// the compression ratio. Values in the range of about 50 to 90 are somewhat reasonable;
/// values above 95 increase the file size with little noticeable improvement to the quality,
/// values below 50 drop the quality substantially.
///
/// This format is well suited for photographs. It is very poorly suited for images with hard
/// edges or text. It does not support transparency.
///
/// JPEG images normally use the `.jpeg` file extension and the `image/jpeg` MIME type.
///
/// See also:
///
///  * <https://en.wikipedia.org/wiki/JPEG>, the Wikipedia page on JPEG.
jpeg,

/// PNG format.
///
/// A loss-less compression format for images. This format is well suited for images
/// with hard edges, such as screenshots or sprites, and images with text. Transparency
/// is supported. The `quality` metric is ignored for this format. The PNG format supports
/// images up to 2,147,483,647 pixels in either dimension, though in practice available
/// memory provides a more immediate limitation on maximum image size.
///
/// PNG images normally use the `.png` file extension and the `image/png` MIME type.
///
/// See also:
///
///  * <https://en.wikipedia.org/wiki/Portable_Network_Graphics>, the Wikipedia page on PNG.
///  * <https://tools.ietf.org/rfc/rfc2083.txt>, the PNG standard.
png,

/// WebP format.
///
/// The WebP format supports both lossy and lossless compression; however, the
/// [Image.toByteData] method always uses lossy compression when [webp] is
/// specified. The `quality` metric determines the compression ratio; higher values
/// result in better quality but larger file sizes, and vice versa. WebP images are
/// limited to 16,383 pixels in each direction (width and height).
///
/// WebP images normally use the `.webp` file extension and the `image/webp` MIME type.
///
/// See also:
///
///  * <https://en.wikipedia.org/wiki/WebP>, the Wikipedia page on WebP.
webp,
```

We should figure out

Hixie commented on testing/dart/encoding_test.dart:
```diff
@@ -0,0 +1,61 @@
+    PictureRecorder recorder = new PictureRecorder();
+    Canvas canvas = new Canvas(recorder, new Rect.fromLTWH(0.0, 0.0, 10.0, 10.0));
+
+    var black = new Paint()
```
rather than using `var`, please specify explicit types

Hixie commented on testing/dart/encoding_test.dart:
```diff
@@ -0,0 +1,61 @@
+
+void main() {
+
+  Image createSquareTestImage() {
```
for clarity, please move this and `readFile` out of `main`

Hixie commented:
This is great! Thanks!

Hixie commented:
Since the method is called toByteData it probably would make sense if it returned a ByteData structure... Not sure how easy it is to get from a Uint8List to a ByteData, off-hand. Maybe @rmacnak-google knows?

tvolkert commented on lib/ui/painting.dart:
```diff
@@ -1215,6 +1225,25 @@ class Image extends NativeFieldWrapperClass2 {
+  ///
+  /// Returns a future which complete with the binary image data (e.g a PNG or JPEG binary data) or
+  /// an error if encoding fails.
+  Future<Uint8List> toByteData({EncodeFormat format: EncodeFormat.JPEG, int quality: 80}) {
```
It feels a little weird to pass quality when it doesn't always make sense depending on the format.  What if the API were as follows:

```dart
/// The encoding formats supported by the [Image.toByteData].
class EncodingFormat {
  const EncodingFormat._(this._format, [this._quality = 0]);

  final int _format;
  final int _quality;

  // Be conservative with the formats we expose. It is easy to add new formats in future but
  // difficult to remove.
  // These values must be kept in sync with the logic in ToSkEncodedImageFormat.
  static const int _jpegFormat = 0;
  static const int _pngFormat = 1;
  static const int _webpFormat = 2;

  /// PNG format.
  ///
  /// A loss-less compression format for images. This format is well suited for images
  /// with hard edges, such as screenshots or sprites, and images with text. Transparency
  /// is supported. The `quality` metric is ignored for this format. The PNG format supports
  /// images up to 2,147,483,647 pixels in either dimension, though in practice available
  /// memory provides a more immediate limitation on maximum image size.
  ///
  /// PNG images normally use the `.png` file extension and the `image/png` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/Portable_Network_Graphics>, the Wikipedia page on PNG.
  ///  * <https://tools.ietf.org/rfc/rfc2083.txt>, the PNG standard.
  static const EncodingFormat png = const EncodingFormat._(_pngFormat);

  /// JPEG format.
  ///
  /// This format, strictly speaking called JFIF, is a lossy compression graphics format that
  /// can handle images up to 65,535 pixels in either dimension. The [quality] metric controls
  /// the compression ratio. Values in the range of about 50 to 90 are somewhat reasonable;
  /// values above 95 increase the file size with little noticeable improvement to the quality,
  /// values below 50 drop the quality substantially.
  ///
  /// This format is well suited for photographs. It is very poorly suited for images with hard
  /// edges or text. It does not support transparency.
  ///
  /// JPEG images normally use the `.jpeg` file extension and the `image/jpeg` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/JPEG>, the Wikipedia page on JPEG.
  factory EncodingFormat.jpeg([int quality = 80]) => new EncodingFormat._(_jpegFormat, quality);

  /// WebP format.
  ///
  /// The WebP format supports both lossy and lossless compression; however, the
  /// [Image.toByteData] method always uses lossy compression when [webp] is
  /// specified. The [quality] metric determines the compression ratio; higher values
  /// result in better quality but larger file sizes, and vice versa. WebP images are
  /// limited to 16,383 pixels in each direction (width and height).
  ///
  /// WebP images normally use the `.webp` file extension and the `image/webp` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/WebP>, the Wikipedia page on WebP.
  factory EncodingFormat.webp([int quality = 80]) => new EncodingFormat._(_webpFormat, quality);
}

Future<Uint8List> toByteData({EncodingFormat format: EncodingFormat.png}) {
  return _futurize(
        (_Callback<Uint8List> callback) => _toByteData(format._format, format._quality, callback)
  );
```

tvolkert replied:
We could even then provide some suggested values...

```dart
  static const EncodingFormat jpegHigh = const EncodingFormat._(_jpegFormat, 95);
  static const EncodingFormat jpegMedium = const EncodingFormat._(_jpegFormat, 80);
  static const EncodingFormat jpegLow = const EncodingFormat._(_jpegFormat, 60);
```

majido replied:
I like this quite a bit more. It provides a much nicer API and enables to have different default quality values for different formats. 

nit: I will probably default the quality to be '100' when unspecified. It is safer if Skia ever decides to respect quality for a format in future.

nit: If we decide to provide additional (Low,Medium, High) suggested values we probably should
do it consistently for all formats where quality is supported.

tvolkert replied:
👍

Hixie replied:
I like this proposal, though please use named arguments not optional arguments for the quality (easier to extend, easier to understand -- a number alone could be any number of things).

We may want to consider making the PNG case a constructor too, so that we can add arguments later.

We don't have to use factory constructors, we could just use real constructors. Then we don't have to have the private constructor.

I think if we're going to have a fallback default, 0 is safer, because it's more likely to cause noticeable problems (like crashes, divide by zero, horrible output, etc), whereas 100 is more likely to cause very subtle problems (like very large file sizes).

Hixie replied:
I probably wouldn't bother with the low/medium/high defaults. I'm not sure what problem they really solve.

majido replied:
Done.  Used a class with three regular constructors, one for each supported format.

tvolkert commented on lib/ui/painting.dart:
```diff
@@ -1215,6 +1289,21 @@ class Image extends NativeFieldWrapperClass2 {
+  ///
+  /// Returns a future which complete with the binary image data (e.g a PNG or JPEG binary data) or
+  /// an error if encoding fails.
+  Future<Uint8List> toByteData({EncodingFormat format: const EncodingFormat.jpeg()}) {
```
... then the test will have to be updated to match too

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1196,6 +1196,80 @@ class Paint {
+  const EncodingFormat.webp({int quality = 80})
+      : _format = _webpFormat,
+        _quality = quality;
+}
```
the constructors should come first

majido replied:
Done.

Hixie commented on lib/ui/painting.dart:
```diff
@@ -1215,6 +1289,24 @@ class Image extends NativeFieldWrapperClass2 {
+  /// an error if encoding fails.
+  Future<ByteData> toByteData({EncodingFormat format: const EncodingFormat.jpeg()}) {
+    return _futurize((_Callback<ByteData> callback) {
+      _toByteData(format._format, format._quality, (Uint8List encoded) {
```
we're discarding the string returned by toByteData here, which seems bad

tvolkert replied:
Good catch.  We just need to `return _toByteData(...` here

majido replied:
Done.

tvolkert commented:
@nsreenath

> Will this also allow us to save a widget tree as PNG ?

https://github.com/flutter/flutter/issues/16274 is the next piece of the puzzle required to do what you ask.

tvolkert commented on lib/ui/painting.dart:
```diff
@@ -1215,6 +1289,24 @@ class Image extends NativeFieldWrapperClass2 {
+  Future<ByteData> toByteData({EncodingFormat format: const EncodingFormat.jpeg()}) {
+    return _futurize((_Callback<ByteData> callback) {
+      _toByteData(format._format, format._quality, (Uint8List encoded) {
+        callback(new ByteData.view(encoded.buffer));
```
Nit: `encoded.buffer.asByteData()` is equivalent and slightly more concise.

majido replied:
Done.

chinmaygarde commented:
Can we land this? The underlying API for accessing the task runners is going to change significantly in https://github.com/flutter/engine/pull/4932 and I want to fix the build after I rebase the contents of this patch. Thanks.

tvolkert commented:
@chinmaygarde, done!  @majido, thanks for the contribution!

nsreenath commented on lib/ui/painting.dart:
```diff
@@ -1215,6 +1289,23 @@ class Image extends NativeFieldWrapperClass2 {
+  ///
+  /// Returns a future which complete with the binary image data (e.g a PNG or JPEG binary data) or
+  /// an error if encoding fails.
+  Future<ByteData> toByteData({EncodingFormat format: const EncodingFormat.jpeg()}) {
```
https://github.com/google/skia/blob/master/src/image/SkImage.cpp#L117

Skia seem to be using lossless PNG format as the default when encoding format is not specified.
Here, are we using JPEG as the default ?
Or should we pass null and use Skia's default ?

tvolkert replied:
Pass `const EncodingFormat.png()` as the format to get lossless PNG encoding.

tvolkert commented:
It looks like this increased our release binary size by ~88K.  Is that expected?

/cc @Hixie @chinmaygarde @majido

jason-simmons commented:
Which binary size increased?

I tried doing android_release builds of libflutter.so before and after this commit and saw a 4KB difference

tvolkert commented:
Hmm, this only seems to substantially affect iOS release binaries.  `complex_layout_ios__compile::release_size_bytes` is the benchmark that regressed, whereas `complex_layout_android__compile::release_size_bytes` only went up by ~4K as you say.

tvolkert commented:
/cc @xster

tvolkert commented:
Looks like the binary size increase is due to the extra dependency on image encoding that we didn't have before.  The reason it only surfaced on iOS and not Android is that we only do LTO on iOS (as a result, Android's binary already had the image encoding support linked in before this change and was thus already inflated).

@majido I think the right answer is to make `Image.toByteData()` return _unencoded_ (raw) bytes (RGBA form, row-primary) and leave it up to callers to encode the raw data however they see fit (using `package:image` or a platform plugin).

This change first landed in `v0.2.10`, meaning that it hasn't hit the `beta` channel yet.  We should make sure we update the API to remove the dependency on image encoding asap so it lands before we cut the next beta release.

tvolkert commented:
Filed https://github.com/flutter/flutter/issues/16537 to fix this.

gedw99 commented:
Do you think that video encinh is possible also at this level , rather than having to write a plug-in ?

tvolkert commented:
@gedw99 see https://github.com/flutter/flutter/issues/8245

gedw99 commented:
@tvolkert thanks !
